### ISSUE TITLE
ActiveRecord marshalling no longer preserves attribute change flags

### ISSUE THREAD

**Issue by tomhughes** [rails/rails#49826]
### Steps to reproduce
The new ActiveRecord marshalling introduced by #47747 does not preserve attribute change flags, as demonstrated by this test case:

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "~> 7.1.1"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.string :title
  end
end

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_marshal_61
    ActiveRecord::Marshalling.format_version = 6.1

    post = Post.new(:title => "Test")
    assert post.title_changed?

    marshalled = Marshal.dump(post)

    unmarshalled = Marshal.load(marshalled)
    assert unmarshalled.title_changed?
  end

  def test_marshal_71
    ActiveRecord::Marshalling.format_version = 7.1

    post = Post.new(:title => "Test")
    assert post.title_changed?

    marshalled = Marshal.dump(post)

    unmarshalled = Marshal.load(marshalled)
    assert unmarshalled.title_changed?
  end
end
```

### Expected behavior
Both tests should pass.

### Actual behavior
Only the test using the old 6.1 marshalling format works.

### System configuration
**Rails version**: 7.1.1

**Ruby version**: 3.2.2

**Comment by fatkodima** [comment]
I am not sure, what is the use case for preserving change flags? The most useful which are already marshalled are just attributes and loaded associations.

cc @byroot

**Comment by tomhughes** [comment]
Well it triggered an issue for as a result of an intersection of two things:

* We have validations on our user model which have a condition of the form `:if => proc { |u| u.email_changed? }` so that they are only enforced when the attribute is changed and not for existing users
* During the signup process we store user records temporarily in the session (which serialises them with `Marshal`) and then retrieve them and save them to the database

The combination of these two things meant that after switching to the new marshalling scheme the validations were not added to the records when they were retrieved from the session because the attribute no longer appeared to be changed.

**Comment by fatkodima** [comment]
> During the signup process we store user records temporarily in the session (which serialises them with Marshal) and then retrieve them and save them to the database

You have a multi step process, right? I would store just a hash with attributes in the session and then create a record at a final stage. I personally feel uncomfortable storing Active Record objects in sessions, caches etc.

**Comment by byroot** [comment]
Yeah, the new format only save the attributes and relations, no other associated states. It's by design to avoid a whole lot of issues, but it true that I didn't deliberately exclude the "dirty attributes" state.

That said it would be a bit tricky to add back, and I really don't recommend relying on it because it's all coupled to the internal representation which we need to change.

**Comment by mguan2020** [comment]
If I'm understanding things correctly, the workaround proposed by @tomhughes seems to solve this issue.

Should the Active Record documentation be updated?

**Comment by bparanj** [comment]
I can document the behavior of 7.1 if this still needs some help. If so, what should be the explanation? Is the following sufficient: 

> ActiveRecord marshalling does not preserves attribute change flags. Do not store user records in the session during signup.
